### PR TITLE
[STORM-1521] When using Kerberos login from keytab with multiple bolts/executors ticket is not renewed

### PR DESCRIPTION
**Pull request by dbahir**
When using a topology that has multiple executors in same process or multiple HBase bolts the tickets do not get renewed. After the ticket expiration the bolts are no longer able to write/read from HBase. That is caused by a login from keytab for every thread, this fixe ensures that there will only be one login per process.

### CODE REVIEW

**Comment by harshach** [comment]
@dbahir patch looks good to me. Can you squash multiple commits into single commit. Thanks

**Comment by harshach** [comment]
@dbahir I think this issue will be in storm-hdfs as well https://github.com/apache/storm/blob/master/external/storm-hdfs/src/main/java/org/apache/storm/hdfs/common/security/HdfsSecurityUtil.java . Can you add changes there as well.

**Comment by dbahir** [comment]
Squashed the commits.

Took a look at the HDFS bolt code path there is a somewhat different and requires additional testing. Created https://issues.apache.org/jira/browse/STORM-1535 to handle it, will create a separate pull request

**Comment by harshach** [comment]
+1. @Parth-Brahmbhatt can you review this as well.

**Comment by Parth-Brahmbhatt** [comment]
@dbahir Does the legacy provider take care of logging in only once internally?

File: external/storm-hbase/src/main/java/org/apache/storm/hbase/security/HBaseSecurityUtil.java [hunk 1]
```diff
@@ -39,26 +39,34 @@
...
+                !(((List) conf.get(TOPOLOGY_AUTO_CREDENTIALS)).contains(AutoHBase.class.getName())))) {
             LOG.info("Logging in using keytab as AutoHBase is not specified for " + TOPOLOGY_AUTO_CREDENTIALS);
-            if (UserGroupInformation.isSecurityEnabled()) {
-                String keytab = (String) conf.get(STORM_KEYTAB_FILE_KEY);
-                if (keytab != null) {
-                    hbaseConfig.set(STORM_KEYTAB_FILE_KEY, keytab);
+            //insure that if keytab is used only one login per process executed
+            if(legacyProvider == null) {
+                synchronized (HBaseSecurityUtil.class) {
+                    if(legacyProvider == null) {
```
Comment (dbahir) [inline external/storm-hbase/src/main/java/org/apache/storm/hbase/security/HBaseSecurityUtil.java]: There is a double check pattern here which locks on the class, the lock on the class will ensure that the legacy provider is instantiated and logged in only once per process

**Comment by Parth-Brahmbhatt** [comment]
Not sure how i missed that you were creating a singleton :-). +1.

**Pull request merged**